src/github/api.rs

```diff
@@ -57,22 +58,155 @@ pub trait Session: Send + Sync {
     fn get_timeline(&self, owner: &str, repo: &str, number: u32) -> Result<Vec<TimelineEvent>>;
 }
 
+
+pub trait GithubSessionFactory: Send + Sync {
+    fn new_session(&self, owner: &str, repo: &str) -> Result<GithubSession>;
+    fn get_token_org(&self, org: &str) -> Result<String>;
+    fn get_token_repo(&self, owner: &str, repo: &str) -> Result<String>;
+    fn bot_name(&self) -> String;
+}
+
+pub fn api_base(host: &str) -> String {
+    if host == "github.com" {
+        "https://api.github.com".to_string()
+    } else {
+        format!("https://{}/api/v3", host)
+    }
+}
+
+pub struct GithubApp {
+    core_remote: Remote,
+    host: String,
+    app_id: u32,
+    // DER formatted API private key
+    app_key: Vec<u8>,
+    app: Option<App>,
+}
+
+pub struct GithubOauthApp {
+    core_remote: Remote,
+    host: String,
+    api_token: String,
+    user: Option<User>,
+}
+
+impl GithubApp {
+    pub fn new(core_remote: Remote, host: &str, app_id: u32, app_key: &[u8]) -> Result<GithubApp> {
+        let mut github = GithubApp {
+            core_remote: core_remote,
+            host: host.into(),
+            app_id: app_id,
+            app_key: app_key.into(),
+            app: None,
+        };
+
+        github.app = Some(github.new_client().get("/app").map_err(|e| {
+            Error::from(format!("Error authenticating to github with token: {}", e))
+        })?);
+
+        info!("Logged in as GithubApp {}", github.bot_name());
+
+        Ok(github)
+    }
+
+    fn new_client(&self) -> HTTPClient {
+        let jwt_token = jwt::new_token(self.app_id, &self.app_key);
+        HTTPClient::new(self.core_remote.clone(), &api_base(&self.host)).with_headers(hashmap!{
+            "Accept" => "application/vnd.github.machine-man-preview+json".to_string(),
```

Comment:
seems a bit spooky? Is this stable for production use?

Reply:
yeah, a bit I suppose. I've found their beta APIs are generally really stable, especially as this one is not brand new.